lib/parsers/mpeg_parser.rb

```diff
@@ -0,0 +1,115 @@
+
+# MPEG Headers documentation:
+# http://dvd.sourceforge.net/dvdinfo/mpeghdrs.html#seq
+# http://www.cs.columbia.edu/~delbert/docs/Dueck%20--%20MPEG-2%20Video%20Transcoding.pdf
+# Useful tool to check the file information: https://www.metadata2go.com/
+class FormatParser::MPEGParser
+  include FormatParser::IOUtils
+
+  MPEG_VALUES = {
+    'mpg' => :mpg,
+    'mpeg' => :mpeg
+  }
+
+  ASPECT_RATIOS = {
+    1 => '1:1',
+    2 => '4:3',
+    3 => '16:9',
+    4 => '2.21:1'
+  }
+
+  FRAME_RATES = {
+    1 => '23.976',
+    2 => '24',
+    3 => '25',
+    4 => '29.97',
+    5 => '30',
+    6 => '50',
+    7 => '59.94',
+    8 => '60'
+  }
+
+  PACK_HEADER_START_CODE = '000001ba'.freeze
+  SEQUENCE_HEADER_START_CODE = 'b3'.freeze
+
+  def likely_match?(filename)
+    filename =~ /\.(mpg|mpeg)$/i
+  end
+
+  def call(io)
+    return unless matches_mpeg_header?(io)
+
+    # We are looping though the stream because there can be several sequence headers and some of them are not usefull.
+    # If we detect that the header is not usefull, then we look for the next one
+    # If we reach the EOF, then the mpg is likely to be corrupted and we return nil
+    loop do
+      search_for_next_sequence_header(io)
+      horizontal_size, vertical_size = parse_image_size(io)
+      radio_hex, rate_hex = parse_rate_information(io)
+
+      if valid_aspect_radio?(radio_hex) && valid_frame_rate?(rate_hex)
+        return file_info(horizontal_size, vertical_size, radio_hex, rate_hex)
+      end
+    end
+  rescue FormatParser::IOUtils::InvalidRead
+    nil
+  end
+
+  def file_info(width_px, height_px, radio_hex_raw, rate_hex_raw)
+    FormatParser::Video.new(
+      format: :mpg,
+      width_px: width_px,
+      height_px: height_px,
```

Comment:
Can we pull framerate up into the main file information struct? The reason is - it seems it will be available always, and we discard the intrinsics when saving metadata because it has a high potential of a) being large and b) including personally identifiable information (EXIF). Same for the aspect ratio. Or will we have to change the struct shape for all video parsers to do so?

Reply:
I thought about this, but if I do that, then all the video parsers will return a null framerate and aspect radio. Do you think that is okay?

Reply:
Good call. Let's keep it in intrinsics then and postpone the decision 👍